packages/server/src/fhir/repo.ts

```diff
@@ -654,14 +649,33 @@ export class Repository extends FhirRepository<PoolClient> implements Disposable
       )?.profile;
       resource.meta = { ...resource.meta, profile: defaultProfiles };
     }
-
     if (!this.supportsInteraction(interaction, resourceType)) {
       throw new OperationOutcomeError(forbidden);
     }
+    return resource;
+  }
+
+  private async updateResourceImpl<T extends Resource>(
+    resource: T,
+    create: boolean,
+    versionId?: string
+  ): Promise<WithId<T>> {
+    const interaction = create ? AccessPolicyInteraction.CREATE : AccessPolicyInteraction.UPDATE;
+    const { resourceType, id } = resource;
+    resource = this.checkResourcePermissions(resource, interaction) as T;
 
-    await preCommitValidation(this.context.projects?.[0], resource, 'update');
+    const preCommitResult = await preCommitValidation(
+      this.context.author,
+      this.context.projects?.[0],
+      resource as WithId<Resource>,
+      'update'
+    );
 
-    const existing = create ? undefined : await this.checkExistingResource<T>(resourceType, id);
+    if (isResourceWithId(preCommitResult, resource.resourceType) && preCommitResult.id === resource.id) {
+      resource = this.checkResourcePermissions(preCommitResult as T, interaction) as T;
```

Comment:
Both of these casts should be removed.

We should make sure that `isResourceWithId` asserts correctly (I think it does).

I think this will be a matter of changing the `resource` parameter type from `T` to `WithId<T>` in both `checkResourcePermissions` and `updateResourceImpl`